### DISCUSSION TITLE
TileCache for custom NavMesh triangulation?

### DISCUSSION THREAD

**Discussion by rob-myers** [isaac-mason/recast-navigation-js#306] [General]
In my use-case I can already construct nice triangulations using [`triangle-wasm`](https://github.com/brunoimbrizi/triangle-wasm), purely in the XZ plane.

Is there any way to skip Recast's triangulation-generation, and construct tiles from a custom triangulation?

It seems natural to want to control the triangulation, either totally, or to some extent.
But I wonder whether I'm barking up the wrong tree, because I don't find any clear references online.

**Comment by isaac-mason** [comment]
This issue tracks supporting this in recast-navigation-js, there hasn't been any progress though: https://github.com/isaac-mason/recast-navigation-js/issues/17

I've also done some searching for common wisdom on how this is done with detour, and haven't found much.

Most have worked around this by running their existing navmesh triangles through recast. This isn't ideal but works fine for many cases.

Using a non-recast navmesh with detour is possible, you'd need to construct detour dtNavMeshCreateParams yourself instead of constructing it with recast. Related: https://github.com/isaac-mason/recast-navigation-js/blob/main/packages/recast-navigation-generators/src/generators/generate-solo-nav-mesh.ts#L353

It's not straightforward though, the data that detour requires is fairly coupled to what recast generates. In particular, recast generates & detour works with both a simple polygon mesh representation, and a detailed mesh representation that contains more accurate height data. This article gives a good overview: https://www.unrealdoc.com/p/navigation-mesh

![image](https://github.com/isaac-mason/recast-navigation-js/assets/67411435/1d87a23d-87eb-472d-947a-7148bbc30503)
![image](https://github.com/isaac-mason/recast-navigation-js/assets/67411435/ed0c5e63-ea68-4cf6-8883-1dd379ac7c59)

I imagine most people wanting to use an existing navmesh with detour already have the detailed mesh representation. Maybe if it's straightforward to generate the simple polygon mesh representation a generic utility could be added to recast-navigation-js

The current recast-navigation-js bindings also don't make it easy to interact with dtNavMeshCreateParams. We could extend DetourNavMeshBuilder to support building navMeshCreateParams without recast structures, maybe using the [Array wrappers](https://github.com/isaac-mason/recast-navigation-js/blob/main/packages/recast-navigation-wasm/src/Arrays.h). Currently there's only methods that take rcPolyMesh and  rcPolyMeshDetail:

```cpp
// https://github.com/isaac-mason/recast-navigation-js/blob/d3adca721ce1993976d4cc01881a61e76e9ecec3/packages/recast-navigation-wasm/src/Detour.cpp#L3

void DetourNavMeshBuilder::setPolyMeshCreateParams(dtNavMeshCreateParams *navMeshCreateParams, rcPolyMesh *polyMesh)
{
    navMeshCreateParams->verts = polyMesh->verts;
    navMeshCreateParams->vertCount = polyMesh->nverts;
    navMeshCreateParams->polys = polyMesh->polys;
    navMeshCreateParams->polyAreas = polyMesh->areas;
    navMeshCreateParams->polyFlags = polyMesh->flags;
    navMeshCreateParams->polyCount = polyMesh->npolys;
    navMeshCreateParams->nvp = polyMesh->nvp;

    rcVcopy(navMeshCreateParams->bmin, polyMesh->bmin);
    rcVcopy(navMeshCreateParams->bmax, polyMesh->bmax);
}

void DetourNavMeshBuilder::setPolyMeshDetailCreateParams(dtNavMeshCreateParams *navMeshCreateParams, rcPolyMeshDetail *polyMeshDetail)
{
    navMeshCreateParams->detailMeshes = polyMeshDetail->meshes;
    navMeshCreateParams->detailVerts = polyMeshDetail->verts;
    navMeshCreateParams->detailVertsCount = polyMeshDetail->nverts;
    navMeshCreateParams->detailTris = polyMeshDetail->tris;
    navMeshCreateParams->detailTriCount = polyMeshDetail->ntris;
}
```

**Reply by rob-myers** [reply]
Thanks for the references.

I originally asked because I wanted certain "doorway polygons" to be present in the recast navmesh.

> <img width="400" alt="constrained-delaunay-triangulation-example" src="https://github.com/isaac-mason/recast-navigation-js/assets/1658537/b2462895-f79e-4b9e-8063-7a41dc5846eb" >

The image above uses a constrained Delaunay triangulation.
Every "doorway" consists of two triangles forming an axis-aligned rectangle.

> <img width="400" alt="recast-triangulation-example" src="https://github.com/isaac-mason/recast-navigation-js/assets/1658537/f40cf8f8-47dd-43f2-b6b4-3ca572728b47">

The image above uses Recast via your wonderful library, where `cs` is `0.149` and `maxSimplificationError` is `0.5`. The indicated doorway does not have this property, and I've failed to fix it by tuning the inputs to your function [`generateTileCache`](https://github.com/isaac-mason/recast-navigation-js/blob/e5708119d09c5b25ddf4243cd1045af6d64a4587/packages/recast-navigation-generators/src/generators/generate-tile-cache.ts#L124).

Given such doorway polygons I can apply filters to prevent agents from navigating through the locked doors. But it's not super-important since this can be achieved via hierarchical path-finding (e.g. rooms are nodes, doors are edges).

So, maybe I'm asking for something which is too complex and not really justified (yet).

**Reply by isaac-mason** [reply]
I think I understand your use case now.

Another way you might be able to support this is by assigning custom areas for the doors as part of the recast navmesh generation process. There's an example of marking areas here: https://recast-navigation-js.isaacmason.com/?path=/story/advanced-custom-areas--compute-path

When marking areas as part of the recast navmesh generation process, you are interacting with the compact heightfield - an intermediate structure before the final navmesh polygons. This means the resulting navmesh polygons are affected by areas.

![image](https://github.com/isaac-mason/recast-navigation-js/assets/67411435/0fb40c83-1f7e-4518-83ee-63c3b4fb2e22)


In the below screenshots from the custom areas storybook, the red area is an area I've marked as unwalkable, and the blue area I've marked as "water".

When I mark a box area as "water", the resulting navmesh polygons wrap around the marked area:

<img width="710" alt="Screenshot 2024-04-10 at 8 50 35 pm" src="https://github.com/isaac-mason/recast-navigation-js/assets/67411435/4483a8b0-a84b-4a51-a6f6-f041adde5dc4">

<img width="706" alt="Screenshot 2024-04-10 at 8 50 43 pm" src="https://github.com/isaac-mason/recast-navigation-js/assets/67411435/1569310e-a37d-4f26-a27b-f9daf0e34056">

**Reply by rob-myers** [reply]
Thanks a lot, this sounds like just what I need. I'll report back.

**Reply by rob-myers** [reply]
It worked!

I kept track of where the doorway triangles were in the input triangulation, then marked them using `markConvexPolyArea`.

Now I can mark the doorways as open/closed, and e.g. avoid agents being pushed through doors by other agents.

**Reply by isaac-mason** [reply]
Glad to hear you got it working! 🙂

**Comment by isaac-mason** [comment]
Related google discussion: https://groups.google.com/g/recastnavigation/c/nrYwdskj8E8/m/UjIHXXkvBAAJ
